src/main/java/eu/dissco/annotationprocessingservice/service/ProcessingService.java

```
@@ -64,43 +71,96 @@ private AnnotationRecord updateExistingAnnotation(AnnotationRecord currentAnnota
     var result = repository.createAnnotationRecord(annotationRecord);
     if (result == SUCCESS) {
       log.info("Annotation: {} has been successfully committed to database", id);
-      var indexDocument = elasticRepository.indexAnnotation(annotationRecord);
-      if (indexDocument.result().jsonValue().equals("updated")) {
+      IndexResponse indexDocument = null;
+      try {
+        indexDocument = elasticRepository.indexAnnotation(annotationRecord);
+      } catch (IOException e) {
+        rollbackUpdatedAnnotation(currentAnnotationRecord, annotationRecord, false);
+      }
+      if (indexDocument.result().equals(Result.Updated)) {
         log.info("Annotation: {} has been successfully indexed", id);
-        kafkaService.publishUpdateEvent(currentAnnotationRecord, annotationRecord);
+        try {
+          kafkaService.publishUpdateEvent(currentAnnotationRecord, annotationRecord);
+        } catch (JsonProcessingException e) {
+          rollbackUpdatedAnnotation(currentAnnotationRecord, annotationRecord, true);
+        }
       }
     }
     return annotationRecord;
   }
 
+  private void rollbackUpdatedAnnotation(AnnotationRecord currentAnnotationRecord,
+      AnnotationRecord annotationRecord, boolean elasticRollback) throws FailedProcessingException {
+    if (elasticRollback) {
+      try {
+        elasticRepository.indexAnnotation(currentAnnotationRecord);
+      } catch (IOException e) {
+        log.error("Fatal exception, unable to rollback update for: {}", annotationRecord, e);
```

You should throw an exception here, or else return null. Otherwise you'll immediately get a NPE when you query `indexDocument.result()`

---

Is this one the right method, this part is a Fatal exception. If we fail to publish a message but then fail to rollback we are a bit stuck and might need manual action to fix this. I don't see where query the `indexDocument.result()`